Implemented initial version of login core functionality and student Total Score display

AndreyWeber opened the pull request:
#2 Implementation.

Stuck on adding blocking wait during user login.

AndreyWeber committed:
Implemented initial version of login core functionality. Should be refactored - wait for async login operation must be added.

AndreyWeber committed:
Implemented login functionality. Must be extended with:
1. Local storage save/load
2. Error handling
3. Login component refactoring
4. Router state handling
5. State refactoring

AndreyWeber committed:
1. Implemented Student Total Score calculation and display.
2. Code refactoring performed

AndreyWeber committed:
Real API Key and Spreadsheet Id values removed from configuration file.

AndreyWeber commented:
Implements #2 and #5

krambertech commented on .eslintrc.js:
```diff
@@ -49,7 +49,6 @@ module.exports = {
         "quotes": ["error", "single", { "avoidEscape": true }],
```
Maybe we need to extend from airbnb config? We will have all latest rules then.

You can apply this:
https://gist.github.com/krambertech/490f27b35d98f7d3954212d55074a547

I'll add a separate task for this

krambertech commented on public/static/config.js:
```diff
@@ -1,5 +1,5 @@
 var APP_CONFIG = {
```
It's better to create sample config than to add placeholders. You can create sample table for that and its id to config

AndreyWeber replied:
Done

krambertech commented on src/actions/index.js:
```diff
@@ -0,0 +1,59 @@
+import {
+    getStudent,
+    getStudentTotalScore
```
We can reduce it on the client. I think its better to get all records and then with the magic of selectors retrieve anything from it

AndreyWeber replied:
Done

krambertech commented on src/actions/index.js:
```diff
@@ -0,0 +1,59 @@
+    userData
+});
+
+export const userLoginFailure = (login, error) => ({
```
No need to have separate actions for everything. You can write dispatch directly in your main action creator

AndreyWeber replied:
Done

krambertech commented on src/api/index.js:
```diff
@@ -1,8 +1,20 @@
 /**
+* Constants
+**/
+const REVIEWS_TAB = 'Reviews';
```
Better to have an object with tabs, not separate variables

AndreyWeber replied:
Done

krambertech commented on src/api/index.js:
```diff
@@ -30,30 +67,15 @@ export function getStudents() {
                 spreadsheetId: config.spreadsheetId,
                 majorDimension: 'ROWS',
-                ranges: ['Students!1:1', 'Students!2:1000'],
+                ranges: [`${STUDENTS_TAB}!1:1`, `${STUDENTS_TAB}!2:1000`],
```
Hardcoded number is bad. Let's move it somewhere in the const

AndreyWeber replied:
Done

krambertech commented on src/selectors/login.js:
```diff
@@ -0,0 +1,13 @@
+
+export const getLogin = state => state.get('login');
+
+export const getLoggedIn = createSelector(
```
isLoggedIn

AndreyWeber replied:
Done

krambertech commented on src/selectors/login.js:
```diff
@@ -0,0 +1,13 @@
+    login => login.get('loggedIn')
+);
+
+export const getLoggingIn = createSelector(
```
isLoggingIn

AndreyWeber replied:
Done

krambertech commented on src/routes.jsx:
```diff
@@ -1,13 +1,22 @@
-        <Route path="*" component={NotFound} />
+    <Route component={App}>
+        <Redirect from="/" to="login" />
+        <Route component={LoginPage} path="login" />
```
Better to write path first. It's easier to scan routes

AndreyWeber replied:
Done

krambertech commented on src/reducers/user.js:
```diff
@@ -0,0 +1,63 @@
+                .set('error', null);
+        }
+
+        case TOTAL_SCORE_REQUEST: {
```
Why do you have everything in one reducer?

AndreyWeber replied:
Done

krambertech commented on src/reducers/login.js:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        case USER_LOGIN_FAILURE: {
+            user(state, action);
```
What is this and why is it here?

AndreyWeber replied:
Done

krambertech commented on src/containers/LoginPage.jsx:
```diff
@@ -1,20 +1,44 @@
+        });
+    }
+
+    handleClick = login => this.props.userLogin(login);
```
Why do you need this handler, it's useless

AndreyWeber replied:
Done

krambertech commented on src/containers/LoginPage.jsx:
```diff
@@ -1,20 +1,44 @@
+    render() {
         return (
-            <h1>Login page</h1>
+            <Login onClick={this.handleClick} />
```
Why onClick? It is better to have properly named callbacks. They need to express what is going on. 

For example in this case it's better to name it onLogin or smthng

AndreyWeber replied:
Done

krambertech commented on src/containers/LoginPage.jsx:
```diff
@@ -1,20 +1,44 @@
-// styles
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.loggingIn) {
+            return;
```
Why do you need it?

AndreyWeber replied:
Strange moment. Should be OK after the next commit. 'if' required to skip route.replace in case if user still logging in.

krambertech commented on src/components/TotalScore.less:
```diff
@@ -0,0 +1,4 @@
+.root {
+    max-width: 700px;
+    margin: 200px auto;
+}
```
Empty line

AndreyWeber replied:
Fixed

krambertech commented:
User needs to login not by its user name but by unique generated key.

krambertech commented:
Also, enter does not work for login

AndreyWeber committed:
1. User 'key' property used as login Id instead of user 'login' property
2. Sample global login has been added

AndreyWeber committed:
1. Collaborator notes from https://github.com/AndreyWeber/fes-react-course-marksboard/pull/6 pull request are taken into consideration
2. Started implementation pre-population of state before application rendering.

AndreyWeber committed:
1. State structure refactored - 'user' entities moved inside 'login' entity
2. Actions and reducers refactored accordingly to point 1.
3. Added util methods required to work with localStorage.
4. Added save of user key to localStorage during user login dispatch after Google spreadsheet api success response.

AndreyWeber committed:
1. Refactored login functionality - support of automatic login if user key exists in localStorage has bee added
2. Added locationState handling on login
3. Login errors handling has been added

AndreyWeber committed:
Implemented authentication with OAuth 2.0 client ID.
NOTE: In real spreadsheet column with marks on "Reviews" tab called "Mark", in the test spreadsheet it is called "Points"

AndreyWeber committed:
1. Added react-bootstrap package.
2. Login component re-written with help of react-bootstrap.

AndreyWeber committed:
#5 #2 Added default spinner.

AndreyWeber commented:
Implements #8

AndreyWeber committed:
UI design slightly changed

AndreyWeber committed:
Fixed error during running npm build command http://stackoverflow.com/questions/38387544/angular-2-webpack-build-from-uglifyjs-syntaxerror-unexpected-token-punc/40818071#40818071

AndreyWeber committed:
1. Re-factoring performed for components, api, selectors and reducers.
2. Cascade styles has been added for Loader, TotalScore and Login pages
3. Backgrounds has been added for Login and TotalScore pages
4. React Bootstrap components used

Implements #8, #5, #2
